marker_api/src/sem/item.rs

```diff
@@ -0,0 +1,160 @@
+use std::marker::PhantomData;
+
+use crate::common::ItemId;
+
+/// The declared visibility of an item or field.
+///
+/// ```
+/// // An item without a visibility
+/// fn moon() {}
+///
+/// // A public item
+/// pub fn sun() {}
+///
+/// // An item with a restricted scope
+/// pub(crate) fn star() {}
+///
+/// pub trait Planet {
+///     // An item without a visibility. But it still has the semantic visibility
+///     // of `pub` as this is inside a trait declaration.
+///     fn mass();
+/// }
+/// ```
+#[repr(C)]
+#[derive(Debug)]
+#[cfg_attr(feature = "driver-api", derive(typed_builder::TypedBuilder))]
+pub struct Visibility<'ast> {
+    #[cfg_attr(feature = "driver-api", builder(setter(skip), default))]
+    _lifetime: PhantomData<&'ast ()>,
+    kind: VisibilityKind,
+}
+
+impl<'ast> Visibility<'ast> {
+    /// Returns `true` if the item is declared as public, without any restrictions.
+    ///
+    /// ```
+    /// // This returns `true`
+    /// pub fn unicorn() {}
+    ///
+    /// // This returns `false`, since the visibility is restricted to a specified path.
+    /// pub(crate) fn giraffe() {}
+    ///
+    /// // This returns `false`, since the visibility is not defined
+    /// fn dragon() {}
+    /// ```
+    ///
+    /// See [`Visibility::is_pub_in_path`] to detect pub declarations with a
+    /// defined path.
+    pub fn is_pub(&self) -> bool {
+        matches!(self.kind, VisibilityKind::Public | VisibilityKind::DefaultPub)
+    }
+
+    /// Returns `true` if the item is declared as `pub(in ..)` with a path in brackets
+    /// that defines the scope, where the item is visible.
+    pub fn is_pub_in_path(&self) -> bool {
+        matches!(self.kind, VisibilityKind::Path(_))
+    }
+
+    /// Returns `true` if the visibility is declared as `pub(crate)`. This is a
+    /// special case of the `pub(<path>)` visibility.
+    ///
+    /// This function checks if the visibility is restricted and the defined path
+    /// belongs to the root module of the crate. Meaning, that this can also be `true`,
+    /// if the visibility uses `pub(super)` to travel up to the crate root.
+    // Ignore, since the `in crate::example_1` path doesn't work for doc tests
```

Comment:
I wonder how this code is layed out during compilation. That's super-unintuitive.
I've checked some metadata:
```
[marker_api/src/sem/item.rs:27] file!() = "marker_api/src/sem/item.rs"
[marker_api/src/sem/item.rs:28] module_path!() = "rust_out"
[marker_api/src/sem/item.rs:8] module_path!() = "rust_out::example_1"
thread 'main' panicked at marker_api/src/sem/item.rs:30:1:
rust_out::main::_doctest_main_marker_api_src_sem_item_rs_64_0::example_1::Baz
```
The panic mentions the `type_name` of a struct defined inside of an `example_1`. I guess rustdoc generates just a single crate for many dotctests if the code example doesn't include an explicit `main` function.

However, if you add an explicit `main()` function the path to `Baz` becomes this:
```
thread 'main' panicked at marker_api/src/sem/item.rs:4:5:
rust_out::example_1::Baz
```

So to remove the `ignore` in this doctest you can add an explicit `# fn main() {}` somewhere here. This will make `crate::example_1` work.

Reply:
That works perfectly, thank you! :heart: